advanced_alchemy/mixins/unique.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from sqlalchemy import ColumnElement, select
+
+if TYPE_CHECKING:
+    from collections.abc import Hashable
+
+    from sqlalchemy.ext.asyncio import AsyncSession
+    from sqlalchemy.ext.asyncio.scoping import async_scoped_session
+    from sqlalchemy.orm import Session
+    from sqlalchemy.orm.scoping import scoped_session
+    from typing_extensions import Self
+
+
+class UniqueMixin:
+    """Mixin for instantiating objects while ensuring uniqueness on some field(s)."""
+
+    @classmethod
+    async def as_unique_async(
+        cls,
+        session: AsyncSession | async_scoped_session[AsyncSession],
+        *args: Any,
+        **kwargs: Any,
+    ) -> Self:
+        """Instantiate and return a unique object within the provided session based on the given arguments.
+
+        If an object with the same unique identifier already exists in the session, it is returned from the cache.
+
+        Args:
+            session (AsyncSession | async_scoped_session[AsyncSession]): SQLAlchemy async session
+            *args (Any): Columns belonging to a table
+            **kwargs (Any): Columns belonging to a table
+
+        Returns:
+            Self: The unique object instance.
+        """
+        key = cls, cls.unique_hash(*args, **kwargs)
+        cache: dict[tuple[type[Self], Hashable], Self] | None = getattr(session, "_unique_cache", None)
+        if cache is None:
+            cache = {}
+            setattr(session, "_unique_cache", cache)
+        if obj := cache.get(key):
+            return obj
+
+        with session.no_autoflush:
+            statement = select(cls).where(cls.unique_filter(*args, **kwargs)).limit(1)
```

Comment:
Instead of `.limit(1)` DYT we should use `.limit(2)` and then use `.one_or_none()` instead of `first()`? This way, if whatever clause is provided by `cls.unique_filter()` doesn't actually limit the result set to a single (or no) row it will be an error as opposed to passing silently.